File: backend/src/main/java/ibee/webapp/todo_app/core/entity/Person.java
package ibee.webapp.todo_app.core.entity;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import ibee.webapp.todo_app.core.entity.person.contactData.address.PersonAddress;
import ibee.webapp.todo_app.core.entity.person.contactData.emailAddress.PersonEmailAddress;
import ibee.webapp.todo_app.core.entity.person.contactData.nationality.PersonCountry;
import ibee.webapp.todo_app.core.entity.person.contactData.phoneNumber.PersonPhoneNumber;
import ibee.webapp.todo_app.core.entity.person.skill.hardSkill.additionlHardSkill.PersonAdditionalHardSkill;
import ibee.webapp.todo_app.core.entity.person.skill.hardSkill.degree.PersonDegree;
import ibee.webapp.todo_app.core.entity.person.skill.hardSkill.professionQualification.PersonProfessionQualification;
import ibee.webapp.todo_app.core.entity.person.skill.softSkill.PersonSoftSkill;
import ibee.webapp.todo_app.core.entity.person.workExperience.PersonWorkExperience;
import ibee.webapp.todo_app.core.repository.baseRepo.IdentifiableEntity;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@AllArgsConstructor
@NoArgsConstructor
@Data
@Builder
@Table(name = "person"
)
public class Person implements IdentifiableEntity<Long>{

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    //@ValidId
    private Long id;

    @Column(name = "social_record_number", nullable = true, length = 10)
    private Integer socialRecordNumber;
    
    @Column(name = "first_name", nullable = false)
    private String firstName;

    @Column(name = "last_name", nullable = false)
    private String lastName;

    //@Temporal(TemporalType.DATE)
    @Column(name = "birth_date", nullable = false)
    private LocalDate birthDate;

    @OneToMany(mappedBy = "person", fetch = FetchType.LAZY, orphanRemoval = true
        //cascade = CascadeType.ALL,
    )
    @Builder.Default
    private Set<PersonCountry> nationalitys = new HashSet<>();

    @OneToMany(
            mappedBy = "person",
            fetch = FetchType.LAZY,
            cascade = CascadeType.PERSIST,
            orphanRemoval = true
    )
    @Builder.Default
    private Set<PersonAddress> addresses = new HashSet<>();

    @OneToMany(
        mappedBy = "person",
        fetch = FetchType.LAZY,
        //cascade = CascadeType.ALL,
        orphanRemoval = true)
    @Builder.Default
    private Set<PersonPhoneNumber> phones = new HashSet<>();

    @OneToMany(
        mappedBy = "person",
        fetch = FetchType.LAZY,
        //cascade = CascadeType.PERSIST,
        orphanRemoval = true
    )
    @Builder.Default
    private Set<PersonEmailAddress> emails = new HashSet<>();

    
     @OneToMany(
        mappedBy = "person",
        fetch = FetchType.LAZY,
        //cascade = CascadeType.ALL,
        orphanRemoval = true
    )
    @Builder.Default
    private Set<PersonDegree> degrees = new HashSet<>();

     @OneToMany(
        mappedBy = "person",
        fetch = FetchType.LAZY,
        //cascade = CascadeType.ALL,
        orphanRemoval = true
    )
    @Builder.Default
    private Set<PersonProfessionQualification> professions = new HashSet<>();


    @OneToMany(
        mappedBy = "person",
        fetch = FetchType.LAZY,
        //cascade = CascadeType.ALL,
        orphanRemoval = true
    )
    @Builder.Default
    private Set<PersonAdditionalHardSkill> additionalSkills = new HashSet<>();

    @OneToMany(
    mappedBy = "person",
    fetch = FetchType.LAZY,
    //cascade = CascadeType.ALL,
    orphanRemoval = true
    )
    @Builder.Default
    private Set<PersonSoftSkill> softSkills = new HashSet<>();

    @OneToMany(
    mappedBy = "person",
    fetch = FetchType.LAZY,
    //cascade = CascadeType.ALL,
    orphanRemoval = true
    )
    @Builder.Default
    private Set<PersonWorkExperience> workExps = new HashSet<>();

}

File: frontend (2)/src/app/types/person.ts
// ==========================================
// BASE & STAMMDATEN ENTITIES
// ==========================================

export interface Country {
    id?: number;
    code: string;       // ISO 3166-1 Alpha-2
    name: string;
    language?: string;
  }
  
  export interface Address {
    id?: number;
    street: string;
    houseNumber: string;
    zipCode: string;
    city: string;
    country: Country;
  }
  
  export interface PhoneNumber {
    id?: number;
    phoneNumber: string; // Ohne führende 0
    countryCode: string; // z.B. "+43"
    fullNumber?: string;
  }
  
  export interface EmailAddress {
    id?: number;
    emailAddress: string;
  }
  
  export interface EducationInstitution {
    id?: number;
    name: string;
    address?: Address;
  }
  
  // ==========================================
  // SKILL HIERARCHIE (HardSkill Inheritance)
  // ==========================================
  
  export interface HardSkill {
    id?: number;
    name: string;   // z.B. "Master of Science", "Führerschein B"
    level?: string;  // z.B. "MSc", "C1", "Meister"
  }
  
  export interface Degree extends HardSkill {
    weight: number;
    preName: boolean;
    postName: boolean;
  }
  
  export interface ProfessionQualification extends HardSkill {
    weight: number;
  }
  
  export interface AdditionalHardSkill extends HardSkill {
    category?: string; // z.B. "LANGUAGE", "IT", "DRIVING"
  }
  
  export interface SoftSkill {
    id?: number;
    name: string;
  }
  
  // ==========================================
  // PERSON JOIN ENTITIES (1:N / N:M Relations)
  // ==========================================
  
  export interface PersonCountry {
    country: Country;
    mainCountry: boolean;
  }
  
  export interface PersonAddress {
    address: Address;
    mainAddress: boolean;
  }
  
  export interface PersonPhoneNumber {
    phoneNumber: PhoneNumber;
    mainNumber: boolean;
  }
  
  export interface PersonEmailAddress {
    emailAddress: EmailAddress;
    mainEmail: boolean;
  }
  
  export interface PersonDegree {
    degree: Degree;
    institution: EducationInstitution;
    startDate: string; // ISO Date String: yyyy-MM-dd
    endDate?: string | null;
    progressInPercent: number;
  }
  
  export interface PersonProfessionQualification {
    professionQualification: ProfessionQualification;
    educationInstitution?: EducationInstitution;
    startDate: string;
    endDate?: string | null;
    certificateNumber?: string;
  }
  
  export interface PersonAdditionalHardSkill {
    additionalHardSkill: AdditionalHardSkill;
  }
  
  export interface PersonSoftSkill {
    softSkill: SoftSkill;
  }
  
  // ==========================================
  // PERSON MAIN ENTITY
  // ==========================================
  
  export interface Person {
    id?: number;
    socialRecordNumber?: number;
    firstName: string;
    lastName: string;
    birthDate: string; // ISO Date String
  
    nationalitys: PersonCountry[];
    addresses: PersonAddress[];
    phones: PersonPhoneNumber[];
    emails: PersonEmailAddress[];
    degrees: PersonDegree[];
    professions: PersonProfessionQualification[];
    additionalSkills: PersonAdditionalHardSkill[];
    softSkills: PersonSoftSkill[];
  }